rosserial_tivac/src/ros_lib/tivac_hardware.h

```diff
@@ -0,0 +1,226 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE 
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF 
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY 
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+//*****************************************************************************
+//
+// Bare minimum hardware resources allocated for rosserials communication.
+// * 2 LEDs if desired
+// * One UART port, interrupt driven transfers
+// * Two RingBuffers of TX_BUFFER_SIZE and RX_BUFFER_SIZE
+// * Systick Interrupt handler
+//
+//*****************************************************************************
+
+#ifndef ROS_LIB_TIVAC_HARDWARE_H
+#define ROS_LIB_TIVAC_HARDWARE_H
+
+extern "C"
+{
+  #include <inc/hw_memmap.h>
+  #include <inc/hw_ints.h>
+  #include <driverlib/gpio.h>
+  #include <driverlib/rom.h>
+  #include <driverlib/rom_map.h>
+  #include <driverlib/systick.h>
+  #include <driverlib/pin_map.h>
+  #include <driverlib/uart.h>
+  #include <utils/ringbuf.h>
+}
+
+#define SYSTICKHZ  1000UL
+
+#ifdef TARGET_IS_TM4C123_RA1
+#define LED1        GPIO_PIN_3  // Green LED
+#define LED2        GPIO_PIN_2  // Blue LED
+#define LED_PORT    GPIO_PORTF_BASE
+#define LED_PERIPH  SYSCTL_PERIPH_GPIOF
+#endif
+
+#ifdef TARGET_IS_TM4C129_RA0
+#define LED1        GPIO_PIN_1  // D1 LED
+#define LED2        GPIO_PIN_0  // D2 LED
+#define LED_PORT    GPIO_PORTN_BASE
+#define LED_PERIPH  SYSCTL_PERIPH_GPION
+#ifndef TM4C129FREQ
+#error "Must define system clock frequency on: TM4C129FREQ"
+#endif
+#endif
+
+#ifndef ROSSERIAL_BAUDRATE
+#define ROSSERIAL_BAUDRATE 57600
+#endif
+
+class TivaCHardware
+{
+  public:
+    TivaCHardware() {}
+
+    void init()
+    {
+#ifdef TARGET_IS_TM4C123_RA1
+      TivaCHardware::ui32SysClkFreq = MAP_SysCtlClockGet();
+#endif
+#ifdef TARGET_IS_TM4C129_RA0
+      TivaCHardware::ui32SysClkFreq = TM4C129FREQ;
+#endif
+
+      // Setup LEDs
+#if defined(LED_HEARTBEAT) || defined(LED_COMM)
+      MAP_SysCtlPeripheralEnable(LED_PERIPH);
+#endif
+#ifdef LED_HEARTBEAT
+      MAP_GPIOPinTypeGPIOOutput(LED_PORT, LED1);
+#endif
+#ifdef LED_COMM
+      MAP_GPIOPinTypeGPIOOutput(LED_PORT, LED2);
+#endif
+
+      // Enable time keeping
+      TivaCHardware::milliseconds = 0;
+      // Set up timer such that it produces one tick for each millisecond
+      SysTickIntRegister(TivaCHardware::SystickIntHandler);
+      MAP_SysTickPeriodSet(TivaCHardware::ui32SysClkFreq/SYSTICKHZ);
+      MAP_SysTickEnable();
+      MAP_SysTickIntEnable();
+
+      // Enable the peripherals for UART0
+      MAP_SysCtlPeripheralEnable(SYSCTL_PERIPH_UART0);
+      MAP_SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOA);
```

Comment:
Which pins? A0 and A1 for UART0?

If you build your project with 'USB' on CMakeLists.txt, it will include a completely different file `tivac_hardware_usb.h'. Which does not use A0 and A1.

I implemented the hardware class on two separate files, UART and USB, instead of having one file with several ifdef sections.